### ISSUE TITLE
Don't show opaque helms

### ISSUE THREAD

**Issue by jomo** [crafatar/crafatar#117]
Laptiar and Minotar, for example, render no helms for Notch and [this guy](https://sessionserver.mojang.com/session/minecraft/profile/ac62cc72ce094fd4901838b92ef4d619) (MassiveChoons).

Notch has an RGB skin with a completely black helm layer.
MassiveChoons ~~has~~ had an RGBA skin with a completely white helm layer, Alpha channel on 100%.

I don't know how Notch actually renders in game, but according to [this guy on twitter](https://twitter.com/CoderTim_/status/587337199915499521), MassiveChoons does not have a white head in game.

-----

Minotar [creates](https://github.com/minotar/minecraft/blob/eadd79738777a1ef61658e1ab9f9ac5fb5f9ed6d/avatars.go#L167-L174) an "Alpha signature" which is the RGBA values of the first pixel. They loop through all pixels in the helm layer and [set the opacity to 0](https://github.com/minotar/imgd/blob/37cf7bd9f5447c82dc174ae874ef02b786c39838/process.go#L353-L362) if the pixel's RGBA matches the "Alpha signature".
Long story short, the first pixel of the image represents transparency.

-----

This seems to be just another way of Mojang driving you crazy with their stupid bullshit, but I guess we'll need to implement this.

**Comment by Jake0oo0** [comment]
The original code that the renders used had a functionality like this, will try to look into it soon.

**Comment by jomo** [comment]
I've tested a bit and it seems like lapis only ignores helms that are covered with a single color, cravatar replaces black (and apparently white) pixels and minotar works as explained above.

I'm unsure how it actually renders in game because it's just not updating for some reason.

**Comment by jomo** [comment]
Alright, I did a few tests.

Minotar seems to be *mostly* doing the right thing, except there seems to be more than 1 color that's hidden.
Maybe all the colors in the top left 8x8 square? Who knows...

[This skin](https://i.imgur.com/1XWmm50.png) didn't show any hat at all, while [this skin](https://i.imgur.com/fntJ8n7.png) was showing both yellow and blue hat.

I'll try to find out more some day, maybe you can also have a look @Jake0oo0 :sunglasses:

-----

Note: Minecraft isn't buggy at all, so you need to flick off all the options in the `Skin customization` menu in game and flick them all back on to see hat updates. Flicking `Hat` alone doesn't seem to help. You'll need to leave and join a server (maybe works with local world, too), but restarting MC is not necessary after changing your skin.

**Comment by jomo** [comment]
Okay, doing some more research..

This is the skin:
![skin](https://i.imgur.com/1yF0JMD.png)

That is this face + hat:
![face](https://i.imgur.com/iiJP0xi.png) + ![hat](https://i.imgur.com/uVK7O3H.png)

This is how it looks in game:
![in game](https://i.imgur.com/bCxOnOj.png)

Quite confusing that neither the black nor the red is showing up ingame, although I definitely had the hat turned on.

----

So far it seems like 
>  Maybe all the colors in the top left 8x8 square?

is the most plausible way, but it definitely needs more research.

**Comment by jomo** [comment]
Alright, I just spent hours figuring out Mojang's shit: Changing 1 pixel, save, upload, log out, log in, repeat. The result is a whole mess.
Here's a common skin map + a "helm map":

![skin map](https://i.imgur.com/ifO3Qnj.png "skin map") + ![helm map](https://i.imgur.com/QIY2Hxq.png "helm map")

The map is the same for < 1.8, steve and alex skins.

If any of the colored pixels in the helm map is transparent, that part - and only that part - will show it's 2nd layer in game. The map for the hat layer actually spans over the whole right arm and the back of the body.

For example, if any of the yellow pixels would be transparent, the right arm's 2nd layer would show up.

AFAIK Mojang doesn't have a slogan yet, may I suggest *"Go figure"*?

**Comment by jomo** [comment]
In terms of implementation, we only need to check any of this if the image has an alpha channel.
i.e. the [PNG color type](http://www.w3.org/TR/PNG/#11IHDR) is 3, 4 or 6.

**Comment by jomo** [comment]
My latest commit has a hotfix that simply skips helms when the skin image has no alpha channel.
That's what I mentioned in my previous comment.
- This does not yet support skins with alpha channel but opaque helms
- Renders need to be fixed separately

**Comment by jomo** [comment]
See also: [MC-83572](https://bugs.mojang.com/browse/MC-83572)

**Comment by jomo** [comment]
Apparently the Windows 10 edition renders skins slightly different.
I think it might actually have the same bug as we do. See [WEB-110](https://bugs.mojang.com/browse/WEB-110).

**Comment by jomo** [comment]
Two bugs have been filed at Mojang about those weird jacket parts: [MC-82464](https://bugs.mojang.com/browse/MC-82464), and [MC-89760](https://bugs.mojang.com/browse/MC-89760) _(reported by me)_

Here is a more useful version of the jacket map, at least more useful for explaining the problem:  
![jacket map](https://i.imgur.com/eQ7kKuN.png)

**Comment by jomo** [comment]
Forgot to fix avatars as well.
